0008-python3-ssh-template.md

```diff
@@ -0,0 +1,83 @@
+- Feature Name: Python 3 Only Salt-SSH
+- Start Date: 2019-05-01
+- SEP Status: Draft
+- SEP PR: https://github.com/saltstack/salt-enhancement-proposals/pull/11
+- Salt Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Since Python 2 will be deprecated in 2020 we need to figure out a solution for Salt-SSH
+to run on machines regardless of the Python version on the targeted system.
+
+# Motivation
+[motivation]: #motivation
+
+Python 2 will be deprecated in 2020 and Salt plans on supporting only Python 3 in the
+Sodium release or later. Due to this deprecation and migration to Python 3 we will
+need to find a specific solution for Salt-SSH to continue to work against hosts
+that still have only Python 2 installed, since by default Salt-SSH only works against
+hosts that have the same python version.
+
+# Design
+[design]: #detailed-design
+
+The design proposal to allow Salt-SSH to work with Python 3, includes multiple options:
+  1. By default, if python versions match we will simply use the system python and Salt-SSH
+     will run as it always has.
+  2. Add a Salt-SSH configuration (for example: pre_flight) to run raw commands before the
+     Salt-SSH command. This will allow a user to create a custom script they can run to install
+     Python 3. It will only run these pre_flight commands if the tarball is not copied over.
+  3. Update the current ssh_ext_alternatives feature in Salt-SSH to automatically attempt to
+     find the necessary python libraries that Salt-SSH uses to copy over and check to see if
+     they are importable. If a user needs additional libraries other than the default they will
+     need to add those manually to the configuration.
+  4. Maintain and build a Python 3 binary that will be copied over with the tarball to be included
+     in the salt-call run. This will also be gated by a config option that a user will need to enable
+     if they always want to use the binary option.
+
+If the python versions do not match when Salt-SSH is run a message will be printed out asking the user
+if they want to copy over the binary, and informing them of the other alternative options pre_flight and
+ssh_ext_alternatives.
+
+
+How are we going to build the Python 3 binary?
+We will build the binary statically for both x86_64 and ARM architecture. We will include the x86_64
+binary by default in the Salt-SSH package. If the ARM architecture is detected we will include a warning
```

Comment:
As long as there is a way to upgrade this, like you can with salt-cloud -u for the bootstrap script, i think this is fine.

Reply:
Great suggestion :) Added here:

https://github.com/saltstack/salt-enhancement-proposals/pull/11/commits/7cda959762ccd3630af1abb34621489170ed5d59